src/py/flwr/server/superlink/driver/serverappio_servicer_test.py

```diff
@@ -54,3 +86,334 @@ def test_raise_if_true() -> None:
+        self, num_transitions: int
+    ) -> None:
+        """Test `PushTaskIns` not successful if RunStatus is not running."""
+        # Prepare
+        node_id = self.state.create_node(ping_interval=30)
+        run_id = self.state.create_run("", "", "", {}, ConfigsRecord())
+        task_ins = create_task_ins(
+            consumer_node_id=node_id, anonymous=False, run_id=run_id
+        )
+
+        self._transition_run_status(run_id, num_transitions)
+
+        # Execute & Assert
+        self._assert_push_task_ins_not_allowed(task_ins, run_id)
+
+    def test_pull_task_res_successful_if_running(self) -> None:
+        """Test `PullTaskRes` success."""
+        # Prepare
+        node_id = self.state.create_node(ping_interval=30)
+        run_id = self.state.create_run("", "", "", {}, ConfigsRecord())
+
+        # Transition status to running. PushTaskRes is only allowed in running status.
+        self._transition_run_status(run_id, 2)
+        request = PullTaskResRequest(node=Node(node_id=node_id), run_id=run_id)
+
+        # Execute
+        response, call = self._pull_task_res.with_call(request=request)
+
+        # Assert
+        assert isinstance(response, PullTaskResResponse)
+        assert grpc.StatusCode.OK == call.code()
+
+    def _assert_pull_task_res_not_allowed(self, node_id: int, run_id: int) -> None:
+        """Assert `PullTaskRes` not allowed."""
+        run_status = self.state.get_run_status({run_id})[run_id]
+        request = PullTaskResRequest(node=Node(node_id=node_id), run_id=run_id)
+
+        with self.assertRaises(grpc.RpcError) as e:
+            self._pull_task_res.with_call(request=request)
+        assert e.exception.code() == grpc.StatusCode.PERMISSION_DENIED
+        assert e.exception.details() == self.status_to_msg[run_status.status]
+
+    @parameterized.expand(
+        [
+            (0,),  # Test not successful if RunStatus is pending.
+            (1,),  # Test not successful if RunStatus is starting.
+            (3,),  # Test not successful if RunStatus is finished.
+        ]
+    )  # type: ignore
+    def test_pull_task_res_not_successful_if_not_running(
+        self, num_transitions: int
+    ) -> None:
+        """Test `PullTaskRes` not successful if RunStatus is not running."""
+        # Prepare
+        node_id = self.state.create_node(ping_interval=30)
+        run_id = self.state.create_run("", "", "", {}, ConfigsRecord())
+
+        self._transition_run_status(run_id, num_transitions)
+
+        # Execute & Assert
+        self._assert_pull_task_res_not_allowed(node_id, run_id)
+
+    def test_push_serverapp_outputs_successful_if_running(self) -> None:
+        """Test `PushServerAppOutputs` success."""
+        # Prepare
+        node_id = self.state.create_node(ping_interval=30)
+        run_id = self.state.create_run("", "", "", {}, ConfigsRecord())
+
+        maker = RecordMaker()
+        context = Context(
+            run_id=run_id,
+            node_id=node_id,
```

Comment:
The ServerApp's node ID (i.e., the node ID of the SuperLink) should be zero, no? I think it's better to set `node_id=0`, and remove the `create_node` line.